lecture-notes/3-bld.md

```diff
@@ -983,17 +985,135 @@ The above diagram with `main.cc` and `square.cc` (C++ files, not C files, but
 same idea) can be compiled two ways, both with the same outcome. The first way
 is as the diagram shows: separate compilation.
 
-```console
+```console?prompt=$
 $ c++ main.cc -c  # compile
 $ c++ square.cc -c  # compile
 $ c++ main.o square.o  # link
 ```
 
+This is the way we have been building programs in our Makefiles throughout
+this module. Now you know what the compilation does but there is still this
+mysterious step that combines object files.
+
+Let's figure out what is going on by experimenting a little bit.
+
+```c
+// main.c
+#include <stdio.h>
+
+extern int random_number();
+
+int main() {
+  printf("Your random number is: %d\n", random_number());
+}
+```
+
+```c
+// lib.c
+int random_number() {
+  return 4;
+}
+```
+
+```console?prompt=$
+$ gcc -Os -c main.c
+$ objdump -M intel --disassemble=main main.o
+0000000000000000 <main>:
+   0:	f3 0f 1e fa          	endbr64
+   4:	50                   	push   rax
+   5:	31 c0                	xor    eax,eax
+   7:	e8 00 00 00 00       	call   c <main+0xc>
+   c:	48 8d 35 00 00 00 00 	lea    rsi,[rip+0x0]        # 13 <main+0x13>
+  13:	bf 01 00 00 00       	mov    edi,0x1
+  18:	89 c2                	mov    edx,eax
+  1a:	31 c0                	xor    eax,eax
+  1c:	e8 00 00 00 00       	call   21 <main+0x21>
+  21:	31 c0                	xor    eax,eax
+  23:	5a                   	pop    rdx
+  24:	c3                   	ret
+$
+```
+
+In particular, it's important to note that the address for the function
+`random_number` is just `0`. You can see this by looking at the `call`
```

Comment:
Think we've gotta make this bold or at least reference the specific line, or else it'll fly over most students' heads.

Reply:
Never mind, you do this right after.